software/common/libs/framing/framing_streams.h

```diff
@@ -0,0 +1,169 @@
+  }
+
+  // accessors for testing purposes
+  size_t Count() { return count_written_; };
+  size_t Flags() { return flags_; };
+
+ protected:
+  // The amount of bytes written
+  size_t count_written_{0};
+  // Status flags
+  uint32_t flags_{0};
+};
+
+// Stream interface
+class Stream {
+ public:
+  // Puts a byte into the stream.
+  // EndOfStream is a special char that designates the end of the stream. We use this for framing.
+  virtual StreamResponse put(int32_t b) = 0;
+};
+
+// OutputStream is connected to some output mechanism that knows how many bytes can be written to
+// it at the moment.
+class OutputStream : public Stream {
+  // Returns number of bytes that can be written.
+ public:
+  virtual size_t bytes_available_for_write() = 0;
+};
+
+#define TO_BYTE(UI) static_cast<uint8_t>((UI)&0x000000FF)
+
+// Passes bytes to the child Stream
+// Calculates CRC while passing bytes
+// Emits CRC to the child Stream upon receiving EndOfStream
+class CrcStream : public Stream {
+ public:
+  explicit CrcStream(Stream &os) : output_(os){};
+
+  StreamResponse put(int32_t b) override {
+    StreamResponse ret;
+    if (EndOfStream == b) {
+      ret = emit_CRC_and_reset();
+      ret += output_.put(EndOfStream);
+      return ret;
+    }
+    crc_ = crc32_single(crc_, static_cast<uint8_t>(b));
+    return output_.put(b);
+  }
+
+ private:
+  StreamResponse emit_CRC_and_reset() {
+    StreamResponse ret;
+    ret = output_.put(TO_BYTE(crc_ >> 24));
+    ret += output_.put(TO_BYTE(crc_ >> 16));
+    ret += output_.put(TO_BYTE(crc_ >> 8));
+    ret += output_.put(TO_BYTE(crc_));
+    crc_ = ResetCRC;
+    return ret;
+  }
+
+  static constexpr uint32_t ResetCRC = 0xFFFFFFFF;
+  Stream &output_;
+  uint32_t crc_{ResetCRC};
+};
+
+// Adds framing markers and escapes special chars
+// Will emit FramingMark:
+// * upon receiving EndOfStream
+// * next time put called after receiving EndOfStream
+// Will emit FramingEscape upon receiving FramingMark or FramingEscape characters
+class EscapeStream : public Stream {
+ public:
+  explicit EscapeStream(Stream &os) : output_(os){};
+
+  StreamResponse put(int32_t b) override {
+    StreamResponse ret = start_frame_if_done();
+    switch (b) {
+      case EndOfStream:
+        ret += output_.put(FramingMark);
+        ret += output_.put(b);
+        frame_done();
+        return ret;
+      case FramingMark:
+      case FramingEscape:
+        ret += output_.put(FramingEscape);
+        ret += output_.put(EscapedChar(b));
+        frame_started_maybe();
+        return ret;
+      default:
+        ret += output_.put(b);
+        frame_started_maybe();
+        return ret;
+    };
+  }
+
+ private:
+  StreamResponse start_frame_if_done() {
+    return is_frame_done_ ? output_.put(FramingMark)
+                          : StreamResponse(0, ResponseFlags::StreamSuccess);
+  }
+
+  void frame_done() {
+    if (!is_frame_done_) {
+      is_frame_done_ = true;
+    }
+  }
+  void frame_started_maybe() {
```

Comment:
I guess that using maybe instills some uncertainty in the reader (and certainly in the current reviewer) maybe add some explanatory comment.